File: backend/app/services/transaction_semantics.py
"""Shared transaction direction and cash-impact semantics.

The transaction direction is interpreted with its payment-method type.  This
matrix is the backend source of truth for API validation, main-bank balances,
and monthly summary aggregates.
"""
from dataclasses import dataclass


@dataclass(frozen=True)
class TransactionSemantics:
    """Effects per €1 of a valid transaction direction.

    ``bank_balance_delta`` affects the linked effective main bank.  Summary
    deltas are reported as positive income and signed net outcomes; a refund
    therefore reduces outcomes.
    """

    bank_balance_delta: int = 0
    income_delta: int = 0
    outcome_delta: int = 0


_PURCHASE = TransactionSemantics(bank_balance_delta=-1, outcome_delta=1)
_REFUND = TransactionSemantics(bank_balance_delta=1, outcome_delta=-1)
_INCOME = TransactionSemantics(bank_balance_delta=1, income_delta=1)
_NON_BANK_PURCHASE = TransactionSemantics(outcome_delta=1)
_NON_BANK_INCOME = TransactionSemantics(income_delta=1)
_CARD_PAYMENT = TransactionSemantics(bank_balance_delta=-1, outcome_delta=1)


TRANSACTION_SEMANTICS_BY_PM_TYPE: dict[str, dict[str, TransactionSemantics]] = {
    "bank": {
        "income": _INCOME,
        "debit": _PURCHASE,
    },
    "debit_card": {
        "debit": _PURCHASE,
        "credit": _REFUND,
    },
    "credit_card": {
        "debit": _PURCHASE,
        "credit": _CARD_PAYMENT,
    },
    "revolving": {
        "debit": _PURCHASE,
        "credit": _CARD_PAYMENT,
    },
    "prepaid": {
        "income": _NON_BANK_INCOME,
        "debit": _NON_BANK_PURCHASE,
    },
    "cash": {
        "income": _NON_BANK_INCOME,
        "debit": _NON_BANK_PURCHASE,
    },
}


def directions_for_payment_method(payment_method_type: str) -> set[str]:
    return set(TRANSACTION_SEMANTICS_BY_PM_TYPE.get(payment_method_type, {}))


def transaction_semantics(
    payment_method_type: str, transaction_direction: str
) -> TransactionSemantics | None:
    return TRANSACTION_SEMANTICS_BY_PM_TYPE.get(payment_method_type, {}).get(transaction_direction)

File: frontend/src/utils/transactionSemantics.ts
import type { PaymentMethod, Transaction } from '../types/api';

type TransactionDirection = Transaction['transaction_direction'];

export interface SupportedTransactionCashImpact {
  isSupported: true;
  direction: TransactionDirection;
  label: string;
  displaySign: '+' | '-';
  bankBalanceDelta: -1 | 0 | 1;
  summaryIncomeDelta: -1 | 0 | 1;
  summaryOutcomeDelta: -1 | 0 | 1;
}

export interface UnsupportedTransactionCashImpact {
  isSupported: false;
  label: 'Unsupported';
  bankBalanceDelta: 0;
  summaryIncomeDelta: 0;
  summaryOutcomeDelta: 0;
}

export type TransactionCashImpact = SupportedTransactionCashImpact | UnsupportedTransactionCashImpact;

const PURCHASE: SupportedTransactionCashImpact = {
  isSupported: true,
  direction: 'debit',
  label: 'Purchase',
  displaySign: '-',
  bankBalanceDelta: -1,
  summaryIncomeDelta: 0,
  summaryOutcomeDelta: 1,
};
const REFUND: SupportedTransactionCashImpact = {
  isSupported: true,
  direction: 'credit',
  label: 'Refund',
  displaySign: '+',
  bankBalanceDelta: 1,
  summaryIncomeDelta: 0,
  summaryOutcomeDelta: -1,
};
const INCOME: SupportedTransactionCashImpact = {
  isSupported: true,
  direction: 'income',
  label: 'Income',
  displaySign: '+',
  bankBalanceDelta: 1,
  summaryIncomeDelta: 1,
  summaryOutcomeDelta: 0,
};
const CARD_PAYMENT: SupportedTransactionCashImpact = {
  isSupported: true,
  direction: 'credit',
  label: 'Card payment',
  displaySign: '-',
  bankBalanceDelta: -1,
  summaryIncomeDelta: 0,
  summaryOutcomeDelta: 1,
};
const NON_BANK_PURCHASE: SupportedTransactionCashImpact = { ...PURCHASE, bankBalanceDelta: 0 };
const NON_BANK_INCOME: SupportedTransactionCashImpact = { ...INCOME, bankBalanceDelta: 0 };
const UNSUPPORTED: UnsupportedTransactionCashImpact = {
  isSupported: false,
  label: 'Unsupported',
  bankBalanceDelta: 0,
  summaryIncomeDelta: 0,
  summaryOutcomeDelta: 0,
};

const CASH_IMPACTS_BY_PAYMENT_METHOD: Record<
  PaymentMethod['type'],
  Partial<Record<TransactionDirection, SupportedTransactionCashImpact>>
> = {
  bank: { debit: PURCHASE, income: INCOME },
  debit_card: { debit: PURCHASE, credit: REFUND },
  credit_card: { debit: PURCHASE, credit: CARD_PAYMENT },
  revolving: { debit: PURCHASE, credit: CARD_PAYMENT },
  prepaid: { debit: NON_BANK_PURCHASE, income: NON_BANK_INCOME },
  cash: { debit: NON_BANK_PURCHASE, income: NON_BANK_INCOME },
};

export function directionsForPaymentMethod(type: PaymentMethod['type'] | undefined): SupportedTransactionCashImpact[] {
  if (!type) return [];
  return Object.values(CASH_IMPACTS_BY_PAYMENT_METHOD[type]).filter(
    (impact): impact is SupportedTransactionCashImpact => impact !== undefined,
  );
}

export function cashImpactForTransaction(
  type: PaymentMethod['type'] | undefined,
  direction: TransactionDirection,
): TransactionCashImpact {
  return (type && CASH_IMPACTS_BY_PAYMENT_METHOD[type][direction]) ?? UNSUPPORTED;
}
